R/Classification.R

```diff
@@ -354,35 +360,51 @@ ScoreCellsWithSavedModel <- function(seuratObj, model, fieldName, batchSize = 20
   colnames(gene_expression_matrix) <- make.names(colnames(gene_expression_matrix))
 
-  # TODO: do all models have this? Maybe mlr3 has some method to pull out features from the classifier instead?
+  # TODO: Is there is more universal way to get the model to report its features?
   # See: https://mlr3.mlr-org.com/reference/LearnerClassif.html
-  modelFeats <- colnames(classifier$model$W)
-  modelFeats <- modelFeats[modelFeats != 'Bias']
-
-  missing <- modelFeats[!modelFeats %in% colnames(gene_expression_matrix)]
-  if (length(missing) > 0) {
-    # TODO: unsure if this is actually a good idea or not?
-    warning(paste0('The following features are used in the model and missing from the input: ', paste0(sort(missing), collapse = ',')))
-    print('Adding zeros for these features')
-    missingMat <- matrix(data = 0, nrow = nrow(gene_expression_matrix), ncol = length(missing))
-    rownames(missingMat) <- rownames(gene_expression_matrix)
-    colnames(missingMat) <- missing
-    gene_expression_matrix <- cbind(gene_expression_matrix, missingMat)
+  modelFeats <- NULL
+  if ('importance' %in% classifier$properties) {
+    modelFeats <- names(classifier$importance())
+  } else if (!is.null(classifier$model) && grepl(x = classifier$model$TypeDetail, pattern = 'logistic regression')) {
+    modelFeats <- colnames(classifier$model$W)
+    toRemove <- names(classifier2$param_set$levels)
+    modelFeats <- modelFeats[!modelFeats %in% toRemove]
+  }
+
+  if (!all(is.null(modelFeats))){
+    missing <- modelFeats[!modelFeats %in% colnames(gene_expression_matrix)]
+    if (length(missing) > 0) {
+      stop(paste0('The following features are used in the model and missing from the input: ', paste0(sort(missing), collapse = ',')))
+    }
+
+    # Subset input data to match model:
+    gene_expression_matrix <- gene_expression_matrix[,colnames(gene_expression_matrix) %in% modelFeats, drop = FALSE]
+  } else {
+    warning(paste0('Unable to infer features from model, type: ', classifier$model$TypeDetail))
   }
-  
-  gene_expression_matrix <- gene_expression_matrix[,colnames(gene_expression_matrix) %in% modelFeats, drop = FALSE]
 
   nBatches <- ifelse(is.na(batchSize), yes = 1, no = ceiling(nrow(gene_expression_matrix) / batchSize))
   probability_vector <- NULL
+  probability_vector_neg <- NULL
   for (batchIdx in 1:nBatches){
     start <- 1 + ((batchIdx-1) * batchSize)
     end <- min((batchIdx * batchSize), nrow(gene_expression_matrix))
     print(paste0("Iteration ", batchIdx, " of ", nBatches, ", (", start, "-", end, ")"))
-    dat <- stats::predict(classifier, newdata = data.frame(gene_expression_matrix[start:end,]), predict_type = 'prob')[,2] #columns are named '0','1', so second column is '1'
+
+    #columns are named '0','1', so the first column is '0' and the second column is '1'
+    dat <- stats::predict(classifier, newdata = data.frame(gene_expression_matrix[start:end,]), predict_type = 'prob')
+
```

Comment:
@bbimber up to you, but in theory we could naively feed a multiclass classifier in here and it won't break. 

something like a `dim(dat)[,2] == 2` check @L396 should do it?

Reply:
good point